platforms/core-runtime/launcher/src/integTest/groovy/org/gradle/launcher/daemon/DaemonToolchainIntegrationTest.groovy

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.launcher.daemon
+
+import org.gradle.api.JavaVersion
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.integtests.fixtures.daemon.DaemonToolchainIntegrationSpec
+import org.gradle.internal.jvm.inspection.JvmVendor
+import org.gradle.internal.os.OperatingSystem
+import org.gradle.test.precondition.Requires
+import org.gradle.test.preconditions.IntegTestPreconditions
+import org.junit.Assume
+
+@Requires(IntegTestPreconditions.NotEmbeddedExecutor)
+class DaemonToolchainIntegrationTest extends DaemonToolchainIntegrationSpec {
+
+    @Requires(IntegTestPreconditions.JavaHomeWithDifferentVersionAvailable)
+    def "Given daemon toolchain version When executing any task Then daemon jvm was set up with expected configuration"() {
+        def otherJvm = AvailableJavaHomes.differentVersion
+        def otherMetadata = AvailableJavaHomes.getJvmInstallationMetadata(otherJvm)
+
+        given:
+        createDaemonJvmToolchainCriteria(otherMetadata.languageVersion.majorVersion)
+
+        expect:
+        succeedsSimpleTaskWithDaemonJvm(otherJvm)
```

Comment:
For example, instead of `createDaemonJvmToolchainCriteria` writing the file itself, I think these tests would be much more "standard" if it just returned a string and the integ test wrote the file contents. Something like this would be more idiomatic:

```
file("gradle/gradle-build.properties") << createDaemonJvmToolchainCriteria(...)
```

This makes it more clear when looking at the test what the build file structure looks like. 

We could even make updates to `AbstractIntegrationSpec` (or perhaps just `DaemonToolchainIntegrationSpec`) to add a `getBuildPropertiesFile` method, similar to the existing [getPropertiesFile](https://github.com/gradle/gradle/blob/993ba9d800aff324d73f4ef69d93d6b230726346/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy#L255-L256) method, which would let us do something like:

```
buildProperties << """
    daemon.jvm.toolchain.version=${otherMetadata.languageVersion.majorVersion}
"""
```

This way, just by looking at the integ test I can see exactly what the contents of the file are without needing to understand whatever the test fixtures are doing.

Reply:
Here depends on the code style, definitely I prefer to keep them as simple as possible since `createDaemonJvmToolchainCriteria` is just creating a build file with some criteria allowing to focus on other aspects of the tests like more complicated scenario/conditions and etc. At the same time, the title of the test already indicates what scenario the tests validates, considering that adding the creation of the file for each test will create to much boilerplate.

I also consider that the suggested code still doesn't improve the readability since the properties created aren't known and you are kind on the same scenario and potentially renaming the function with `createBuildPropertiesWithDaemonToolchainCriteria(...)` will make easier to understand what the function is doing.
```
buildProperties << createDaemonJvmToolchainCriteria(...)
```

Having said that, I'm happy to change those with however you consider the best following the Gradle standards, and keeping this in mind to apply for future tests.

Reply:
I went into detail here since you specifically asked for comments on style/structure. I probably wouldn't have otherwise, but from experience I see a lot more files, and have a better time understanding files, that make the contents of written files very clear from looking at the test method body. 

We should not rely on the test name adding context to the test content -- I've seen plenty of tests where the test name no longer reflects the actual content of the tests. And while that may not be great, it does happen. I would much prefer that the functionality of the test is clear by looking at the actual content of the body as opposed to needing to read the method name. 

Personally, I see the `createDaemonJvmToolchainCriteria` method as very opaque. It puts a lot of importance on the ordering of parameters, and while it may work well currently, as there are only three fields in the file, what happens if there are more? I do not see this fixture as very extensible. 

I don't think we need to revisit all this now -- as long as the tests do what they say they do. 

But, I think it would be better for future tests to be written where file contents test expectations are more clear by looking at the method body.